core_v14/src/substrate_type.rs

```diff
@@ -1,275 +1,352 @@
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str("Compact<")?;
+        Debug::fmt(&*self.value, f)?;
+        f.write_str(">")
+    }
+}
 
-	s.push('(');
-	for substring in fields.iter() {
-		s.push_str(&substring.to_string());
-        s.push_str(", ");
-	}
-	s.push(')');
-	s
+/// Represent a `bitvec`.
+pub struct BitSequenceType {
+    /// What order do we read the bits from the underlying type.
+    pub bit_order_type: Box<SubstrateType>,
+    /// I believe this is what the underlying storage looks like; eg are
+    /// the bits packed into u8's or u16's etc.
+    pub bit_store_type: Box<SubstrateType>
 }
 
-impl Display for SubstrateType {
-	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-		match self {
-			SubstrateType::TypePointer(t) => write!(f, "{}", t),
-			SubstrateType::Unit(u) => write!(f, "{}", u),
-			SubstrateType::Struct(t) => {
-				for substring in t.iter() {
-					write!(f, "{}, ", substring)?;
-				}
-                Ok(())
-			}
-			SubstrateType::Set(t) => {
-				for substring in t.iter() {
-					write!(f, "{}, ", substring)?;
-				}
-                Ok(())
-			}
-			SubstrateType::Tuple(t) => write!(f, "{}", &display_types(t)),
-			SubstrateType::Enum(t) => {
-				write!(f, "{}", "{ ");
-				for field in t.iter() {
-					write!(f, "{} ,", field.to_string())?;
-				}
-				write!(f, "{}", " }")
-			}
-			SubstrateType::Array { size, ty } => write!(f, "{}", &format!("[{};{}], ", ty, size)),
-			SubstrateType::Std(t) => write!(f, "{}", t.to_string()),
-			SubstrateType::Generic(outer, inner) => write!(f, "{}", &format!("{}<{}>", outer, inner)),
-			SubstrateType::Number => write!(f, "number"),
-			SubstrateType::U8 => write!(f, "u8"),
-			SubstrateType::U16 => write!(f, "u16"),
-			SubstrateType::U32 => write!(f, "u32"),
-			SubstrateType::U64 => write!(f, "u64"),
-			SubstrateType::U128 => write!(f, "u128"),
-			SubstrateType::USize => write!(f, "usize"),
+impl Debug for BitSequenceType {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str("BitSequence")
+    }
+}
 
-			SubstrateType::I8 => write!(f, "i8"),
-			SubstrateType::I16 => write!(f, "i16"),
-			SubstrateType::I32 => write!(f, "i32"),
-			SubstrateType::I64 => write!(f, "i64"),
-			SubstrateType::I128 => write!(f, "i128"),
-			SubstrateType::ISize => write!(f, "isize"),
+/// This represents an error when attempting to convert from a [`scale_info::Type`] to a [`SubstrateType`].
+#[derive(Debug, Clone, thiserror::Error)]
+pub enum ConvertError {
+    #[error("could not find type with ID {0} in the type registry")]
+    TypeNotFound(u32)
+}
 
-			SubstrateType::F32 => write!(f, "f32"),
-			SubstrateType::F64 => write!(f, "f64"),
+impl SubstrateType {
+    /// Convert a type from a [`ScaleInfoType`] to a [`SubstrateType`]. We do this because we can't manually
+    /// construct and work with some of the [`ScaleInfoType`] values, but we'd like to be able to do so to
```

Comment:
Couldn't we make it possible to change `scale-info` to make such manual construction possible?